codecs/resize/src/lib.rs

```diff
@@ -77,44 +83,64 @@ pub fn resize(
     let num_input_pixels = input_width * input_height;
     let num_output_pixels = output_width * output_height;
 
-    let (to_linear, to_color_space) = converter_funcs(color_space_conversion);
-    let (premultiplier, demultiplier) = alpha_multiplier_funcs(premultiply);
+    let mut output_image: Vec<u8> = vec_with_len(num_output_pixels * 4, 0);
 
-    // If both options are false, there is no preprocessing on the pixel valus
+    // If both options are false, there is no preprocessing on the pixel values
     // and we can skip the loop.
-    if premultiply || color_space_conversion {
-        for i in 0..num_input_pixels {
-            for j in 0..3 {
-                input_image[4 * i + j] =
-                    premultiplier(to_linear(input_image[4 * i + j]), input_image[4 * i + 3]);
-            }
+    if !premultiply && !color_space_conversion {
+        let mut resizer = resize::new(
+            input_width,
+            input_height,
+            output_width,
+            output_height,
+            Pixel::RGBA,
+            typ,
+        );
+        resizer.resize(input_image.as_slice(), output_image.as_mut_slice());
+        return output_image;
+    }
+
+    // Otherwise, we convert to f32 images to keep the
+    // conversions as lossless and high-fidelity as possible.
+    let (to_linear, to_srgb) = srgb_converter_funcs(color_space_conversion);
+    let (premultiplier, demultiplier) = alpha_multiplier_funcs(premultiply);
+
+    let mut preprocessed_input_image: Vec<f32> = Vec::with_capacity(input_image.len());
+    preprocessed_input_image.resize(input_image.len(), 0.0f32);
+    for i in 0..num_input_pixels {
+        for j in 0..3 {
+            preprocessed_input_image[4 * i + j] = premultiplier(
+                to_linear(input_image[4 * i + j]),
+                (input_image[4 * i + 3] as f32) / 255.0,
+            );
+            preprocessed_input_image[4 * i + 3] = (input_image[4 * i + 3] as f32) / 255.0;
```

Comment:
This doesn't use `j`, is it meant to be happening inside of the inner `j` loop?

Reply:
Nope. `j` loops over the RGB channel indices. This one sets the alpha channel.

Reply:
oh, I just understood what you mean lol. Yes, gonna move it.